plugin/evm/atomic_trie_test.go

```diff
@@ -591,6 +597,187 @@ func TestApplyToSharedMemory(t *testing.T) {
 	}
 }
 
+func Test_atomicTrie_AcceptTrie(t *testing.T) {
+	t.Parallel()
+
+	metadataHexPrefix := hex.EncodeToString(prefixdb.MakePrefix(atomicTrieMetaDBPrefix))
+
+	testCases := map[string]struct {
+		lastAcceptedRoot        common.Hash
+		lastCommittedRoot       common.Hash
+		lastCommittedHeight     uint64
+		commitInterval          uint64
+		height                  uint64
+		root                    common.Hash
+		wantHasCommitted        bool
+		wantErr                 string
+		wantLastCommittedHeight uint64
+		wantLastCommittedRoot   common.Hash
+		wantLastAcceptedRoot    common.Hash
+		wantTipBufferRoot       common.Hash
+		wantKeyValuePairs       map[string]string
+	}{
+		"no_committing": {
+			lastAcceptedRoot:        types.EmptyRootHash,
+			lastCommittedRoot:       common.Hash{2},
+			lastCommittedHeight:     100,
+			commitInterval:          10,
+			height:                  105,
+			root:                    common.Hash{3},
+			wantLastCommittedHeight: 100,
+			wantLastCommittedRoot:   common.Hash{2},
+			wantLastAcceptedRoot:    common.Hash{3},
+			wantTipBufferRoot:       common.Hash{3},
+			wantKeyValuePairs: map[string]string{
+				metadataHexPrefix + "0000000000000064":// height 100
+				"0200000000000000000000000000000000000000000000000000000000000000",
+				metadataHexPrefix + hex.EncodeToString(lastCommittedKey): "0000000000000064", // height 100
+			},
+		},
+		"no_committing_with_previous_root": {
+			lastAcceptedRoot:        common.Hash{1},
+			lastCommittedRoot:       common.Hash{2},
+			lastCommittedHeight:     100,
+			commitInterval:          10,
+			height:                  105,
+			root:                    common.Hash{3},
+			wantLastCommittedHeight: 100,
+			wantLastCommittedRoot:   common.Hash{2},
+			wantLastAcceptedRoot:    common.Hash{3},
+			wantTipBufferRoot:       common.Hash{3},
+			wantKeyValuePairs: map[string]string{
+				metadataHexPrefix + "0000000000000064":// height 100
+				"0200000000000000000000000000000000000000000000000000000000000000",
+				metadataHexPrefix + hex.EncodeToString(lastCommittedKey): "0000000000000064", // height 100
+			},
+		},
+		"commit_all_up_to_height_without_height": {
+			lastAcceptedRoot:        types.EmptyRootHash,
+			lastCommittedRoot:       common.Hash{2},
+			lastCommittedHeight:     60,
+			commitInterval:          10,
+			height:                  105,
+			root:                    common.Hash{3},
+			wantHasCommitted:        true,
+			wantLastCommittedHeight: 100,
+			wantLastCommittedRoot:   types.EmptyRootHash,
+			wantLastAcceptedRoot:    common.Hash{3},
+			wantTipBufferRoot:       common.Hash{3},
+			wantKeyValuePairs: map[string]string{
+				metadataHexPrefix + "000000000000003c":// height 60
+				"0200000000000000000000000000000000000000000000000000000000000000",
```

Comment:
nit: should we use `hex.EncodeToString(common.Hash{2})` here (seems shorter)?

yet better, should we have `wantPersistedHeightMap map[uint64]common.Hash`?
We can perform the string to hex in the test (if you care about the failure message) or we can call string() on the bytes before using them as a map key and avoid encoding to hex entirely.

the lastCommittedKey should always match `wantLastCommittedHeight` so we can add it to the expected map in the test.

Reply:
Changed those strings to i.e. `hex.EncodeToString(common.Hash{2}.Bytes()),`

Reply:
- I would rather keep this dumb in the test case definition, listing all keys, than adding more logic code in the subtest.

> the lastCommittedKey should always match wantLastCommittedHeight so we can add it to the expected map in the test.

I don't think that it is, there is a case `commit_all_up_to_height_without_height` where the last committed height is `60`, the height `105` and the `wantLastCommittedHeight` is `100` (due to commit interval)